File: bindings/python/examples/headless_stream_test.py
"""
Myth Engine — Headless ReadbackStream Test (Async Ring Buffer)
===============================================================

Acceptance Test B: Renders 100 frames using a ``ReadbackStream`` with
``buffer_count=3``. Frames are submitted non-blocking and collected via
``try_recv()``. Remaining in-flight frames are drained with ``flush()``.
Asserts exactly 100 frames are received.

Usage:
    maturin develop --release
    python examples/headless_stream_test.py
"""

import myth

renderer = myth.Renderer(render_path="basic")
renderer.init_headless(256, 256)

# Minimal scene
scene = renderer.create_scene()
cube = scene.add_mesh(myth.BoxGeometry(1, 1, 1), myth.UnlitMaterial(color=[1.0, 0.4, 0.1]))
cam = scene.add_camera(myth.PerspectiveCamera(fov=45, near=0.1))
cam.position = [0, 2, 5]
cam.look_at([0, 0, 0])
scene.active_camera = cam
scene.add_light(myth.DirectionalLight(color=[1, 1, 1], intensity=3))

TOTAL_FRAMES = 100
EXPECTED_BYTES = 256 * 256 * 4  # RGBA8

stream = renderer.create_readback_stream(buffer_count=3)

received = 0
sent = 0
lost = 0

for i in range(TOTAL_FRAMES):
    renderer.update(1.0 / 60.0)
    renderer.render()

    try:
        stream.try_submit(renderer)
        sent += 1
    except Exception as e:
        lost += 1
        print(f"Error submitting frame {i}: {e}")

    renderer.poll_device()

    # Opportunistically pull ready frames.
    while True:
        frame = stream.try_recv()
        if frame is None:
            break
        assert len(frame["pixels"]) == EXPECTED_BYTES, (
            f"frame {frame['frame_index']}: unexpected size {len(frame['pixels'])}"
        )
        received += 1

# Drain remaining in-flight frames.
remaining = stream.flush(renderer)
for frame in remaining:
    assert len(frame["pixels"]) == EXPECTED_BYTES
    received += 1

assert received == sent, f"expected {sent} frames, got {received}"

print(
    f"Test B passed: Total {TOTAL_FRAMES} frames , sent {sent}, received {received}, lost {lost} "
    f"(256×256, buffer_count=3, {EXPECTED_BYTES} bytes/frame)"
)

renderer.dispose()

File: examples/headless_stream_test.rs
//! Acceptance Test B — `ReadbackStream` (async ring buffer).
//!
//! Renders 100 frames in headless mode using a `ReadbackStream` with
//! `buffer_count = 3`. Frames are submitted non-blocking and collected
//! via `try_recv`. Any remaining in-flight frames are drained with
//! `flush` at the end. Verifies exactly 100 frames are received.
//!
//! ```bash
//! cargo run --example headless_stream_test --no-default-features
//! ```

use myth::prelude::*;
use myth::render::core::ReadbackStream;

fn main() {
    env_logger::init();

    let mut engine = Engine::default();

    let width: u32 = 256;
    let height: u32 = 256;
    pollster::block_on(engine.init_headless(width, height, None)).expect("headless init failed");

    // ── Minimal scene ────────────────────────────────────────────────
    let scene = engine.scene_manager.create_active();

    let mat = UnlitMaterial::new(Vec4::new(1.0, 0.4, 0.1, 1.0));
    let _cube = scene.spawn_box(1.0, 1.0, 1.0, mat, &engine.assets);

    let cam = scene.add_camera(Camera::new_perspective(
        45.0,
        width as f32 / height as f32,
        0.1,
    ));
    scene
        .node(&cam)
        .set_position(0.0, 2.0, 5.0)
        .look_at(Vec3::ZERO);
    scene.active_camera = Some(cam);

    scene.add_light(Light::new_directional(Vec3::ONE, 3.0));

    // ── Create ReadbackStream ────────────────────────────────────────
    let mut stream: ReadbackStream = engine
        .create_readback_stream(3)
        .expect("create_readback_stream failed");

    // Clone device/queue handles (in wgpu 29 they are internally Arc-backed,
    // so cloning is cheap and avoids borrow conflicts with engine.update()).
    let device = engine.device().expect("device").clone();
    let queue = engine.queue().expect("queue").clone();

    let total_frames: u64 = 100;
    let expected_bytes = (width * height * 4) as usize;
    let mut received: u64 = 0;

    // ── Hot loop ─────────────────────────────────────────────────────
    for _ in 0..total_frames {
        engine.update(1.0 / 60.0);
        engine.render_active_scene();

        let texture = engine.headless_texture().expect("headless texture");
        stream
            .submit(&device, &queue, texture)
            .expect("submit failed");

        // Drive GPU callbacks.
        engine.poll_device();

        // Opportunistically pull ready frames.
        while let Some(frame) = stream.try_recv().expect("try_recv failed") {
            assert_eq!(
                frame.pixels.len(),
                expected_bytes,
                "frame {}: unexpected pixel buffer size",
                frame.frame_index
            );
            received += 1;
        }
    }

    // ── Flush remaining ──────────────────────────────────────────────
    stream
        .flush(&device, |frame| {
            assert_eq!(
                frame.pixels.len(),
                expected_bytes,
                "flush frame {}: unexpected pixel buffer size",
                frame.frame_index
            );
            received += 1;
        })
        .expect("flush failed");

    assert_eq!(
        received, total_frames,
        "expected {total_frames} frames, got {received}"
    );

    println!(
        "Test B passed: {total_frames} async readback frames OK \
         ({width}×{height}, buffer_count=3, {expected_bytes} bytes/frame)"
    );
}
